File: zk-castvote/groth16/parameters.go
package groth16

import "fmt"
import "github.com/datachainlab/go-risc0-verifier/risc0"

// selector -> verifier parameters
var risc0SelectorVerifierParameters map[[4]byte]risc0.VerifierParameters

func init() {
	verifierParams := risc0.GetVerifierParameters()
	risc0SelectorVerifierParameters = make(map[[4]byte]risc0.VerifierParameters)
	for _, params := range verifierParams {
		selector := CalculateSelector(params)
		risc0SelectorVerifierParameters[selector] = params
	}
}

// CalculateSelector calculates the selector from the verifier parameters.
func CalculateSelector(params risc0.VerifierParameters) [4]byte {
	var data [130]byte
	copy(data[:32], sha256Bytes([]byte("risc0.Groth16ReceiptVerifierParameters")))
	copy(data[32:64], params.ControlRoot[:])
	copy(data[64:96], params.BN254ControlID[:])
	copy(data[96:128], vkDigest[:])
	copy(data[128:], []byte{0x03, 0x00})
	h := Sha256(data[:])
	var selector [4]byte
	copy(selector[:], h[:4])
	fmt.Println("selector: ", selector)
	return selector
}

// GetVerifierParameters returns the verifier parameters corresponding to the given selector.
func GetVerifierParameters(selector [4]byte) (risc0.VerifierParameters, bool) {
	params, ok := risc0SelectorVerifierParameters[selector]
	return params, ok
}

// GetVerifierParameters2 returns the verifier parameters corresponding to the given selector.
func GetVerifierParameters2(selector []byte) (risc0.VerifierParameters, bool) {
	var s [4]byte
	copy(s[:], selector)
	return GetVerifierParameters(s)
}

File: zk-castvote/python/groth16/parameters.py
import hashlib
from typing import Optional, Dict
from risc0.risc0 import VerifierParameters, get_verifier_parameters as risc0_get_verifier_parameters
from .vk import vk_digest
from .utils import sha256, sha256_bytes


# selector -> verifier parameters
_risc0_selector_verifier_parameters: Dict[bytes, VerifierParameters] = {}


def _init_selector_parameters():
    """Initialize selector to verifier parameters mapping."""
    global _risc0_selector_verifier_parameters
    verifier_params = risc0_get_verifier_parameters()  # Call the risc0 function (no args)
    _risc0_selector_verifier_parameters = {}
    for params in verifier_params.values():
        selector = calculate_selector(params)
        _risc0_selector_verifier_parameters[selector] = params


def calculate_selector(params: VerifierParameters) -> bytes:
    """Calculate the selector from the verifier parameters."""
    data = bytearray(130)
    tag = sha256_bytes(b"risc0.Groth16ReceiptVerifierParameters")
    data[0:32] = tag
    data[32:64] = params.control_root
    data[64:96] = params.bn254_control_id
    data[96:128] = vk_digest
    data[128:130] = bytes([0x03, 0x00])
    h = sha256(bytes(data))
    return h[:4]


def get_verifier_parameters(selector: bytes) -> Optional[VerifierParameters]:
    """Get verifier parameters corresponding to the given selector."""
    if len(_risc0_selector_verifier_parameters) == 0:
        _init_selector_parameters()
    
    if len(selector) < 4:
        return None
    
    selector_4 = selector[:4]
    return _risc0_selector_verifier_parameters.get(selector_4)


def get_verifier_parameters2(selector: bytes) -> Optional[VerifierParameters]:
    """Get verifier parameters corresponding to the given selector (flexible length)."""
    return get_verifier_parameters(selector)

